### DISCUSSION TITLE
How to show tmux's session name inside wezterm tab tiltle?

### DISCUSSION THREAD

**Discussion by younger-1** [wezterm/wezterm#3639] [Q&A]
Basiclly, I want to refresh the tmux's session name inside wezterm tab title when the tmux session changed in the current focused tab. The basic logic is to set a user var for tmux session name and use `set-hook -g client-session-changed` to auto-trigger updating of the var. 

I have read carefully of these doc and tried with following code but it doesn't work. 
Is there a good way to do this?
1: https://wezfurlong.org/wezterm/shell-integration.html
2: https://wezfurlong.org/wezterm/recipes/passing-data.html
3: https://github.com/wez/wezterm/blob/main/assets/shell-integration/wezterm.sh

```lua
wezterm.on('format-tab-title', function(tab, tabs, panes, config, hover, max_width)
  local title = tab.active_pane.title

  local tmux = ''
  local tmux_session = tab.active_pane.user_vars.tmux_panel_session
  local tmux_program = tab.active_pane.user_vars.WEZTERM_PROG
  if tmux_session and tmux_session ~= '' then
    tmux = tmux_session
  end
  if tmux_program and tmux_program ~= '' then
    tmux = tmux .. ':' .. tmux_program
  end
  if #tmux > 0 then
    tmux = tmux .. ' '
  end

  title = title .. tmux

  if tab.is_active then
    return {
      { Background = { Color = 'blue' } },
      { Text = ' ' .. title .. ' ' },
    }
  end
  return title
end)
```

The main pain point is I could NOT get tmux send OSC 1337 Escape sequences to set user vars by **keybind** or by **hook**, to achieve that I have to **manually** input the OSC 1337 Escape sequences in terminal:

```bash
# [1] By keybind:
bind-key M run 'printf "\033Ptmux;\033\033]1337;SetUserVar=%s=%s\007\033\\" "tmux_panel_session" `echo -n "named-session" | base64`'

# [2] By hook: tmux set-hook client-session-changed
if '[ $WEZTERM_PANE ]' {
  # set -p @t_p_s '#S'
  # Ref: https://unix.stackexchange.com/questions/628053/use-bash-functions-in-tmux-run-shell-command
  # set-hook -g client-session-changed 'run -b "__wezterm_set_user_var tmux_panel_session #{E:#S}"'
  set-hook -g client-session-changed 'run -b "cut -c3- ~/.tmux.conf | bash -s __wezterm_set_user_var tmux_panel_session  #{E:#S}"'
}

# [3] Manually
printf "\033Ptmux;\033\033]1337;SetUserVar=%s=%s\007\033\\" "tmux_panel_session" `echo -n "my-session" | base64` at
```

**Comment by wez** [comment]
Did you see this part of https://wezfurlong.org/wezterm/recipes/passing-data.html#user-vars ?

```
      # <https://github.com/tmux/tmux/wiki/FAQ#what-is-the-passthrough-escape-sequence-and-how-do-i-use-it>
      # Note that you ALSO need to add "set -g allow-passthrough on" to your tmux.conf
```

**Reply by younger-1** [reply]
Yes, I had set this option and I can change the tab title by manually inputting the escape sequence
```bash
printf "\033Ptmux;\033\033]1337;SetUserVar=%s=%s\007\033\\" "tmux_panel_session" `echo -n "my-session" | base64` at
```

**Comment by younger-1** [comment]
Update:
I managed to get tmux send OSC 1337 Escape sequences to by **keybind**, using `new-window`
```
bind M new-window 'printf "\033Ptmux;\033\033]1337;SetUserVar=%s=%s\007\033\\" "tmux_panel_session" $(echo -n $(tmux display -p "#S:#W") | base64)'
```

**Comment by younger-1** [comment]
I still can't get around of using hook to refresh session name. Its too hard to me now.

**Reply by LinoWhy** [reply]
tmux native supports [this feature](https://superuser.com/questions/1098077/how-to-keep-terminal-window-title-in-sync-with-tmux-window)
